rs/tests/networking/canister_http_correctness_test.rs

```diff
@@ -610,30 +814,1006 @@ pub fn test_http_calls_to_ic_fails(env: TestEnv) {
+        },
+    ));
+
+    assert_matches!(
+        response,
+        Err(RejectResponse {
+            reject_code: RejectCode::SysFatal,
+            ..
+        })
+    );
+}
+
+fn reference_transform_function_exposed_by_different_canister(env: TestEnv) {
+    let handlers = Handlers::new(&env);
+    let webserver_ipv6 = get_universal_vm_address(&env);
+    let url = format!(
+        "https://[{}]:20443/{}/{}",
+        webserver_ipv6, "ascii", "hello_world"
+    );
+
+    let proxy_canister_id_1 = get_proxy_canister_id(&env);
+    // Create another proxy canister;
+    // Get application subnet node to deploy canister to.
+    let mut nodes = get_node_snapshots(&env);
+    let node = nodes.next().expect("there is no application node");
+    let runtime = get_runtime_from_node(&node);
+    let _ = create_proxy_canister(&env, &runtime, &node);
+    let proxy_canister_id_2 = get_proxy_canister_id(&env);
+
+    assert_ne!(
+        proxy_canister_id_1, proxy_canister_id_2,
+        "create_proxy_canister() should create a new proxy canister with a new canister id."
+    );
+
+    let request = CanisterHttpRequestArgs {
+        url,
+        headers: BoundedHttpHeaders::new(vec![]),
+        method: HttpMethod::GET,
+        body: Some("".as_bytes().to_vec()),
+        max_response_bytes: None,
+        transform: Some(TransformContext {
+            function: TransformFunc(candid::Func {
+                principal: proxy_canister_id_1.into(),
+                method: "test_transform".to_string(),
+            }),
+            context: vec![],
+        }),
+    };
+
+    let response = block_on(submit_outcall(
+        &handlers,
+        RemoteHttpRequest {
+            request: request.clone(),
+            cycles: 500_000_000_000,
+        },
+    ));
+
+    assert_matches!(
+        response,
+        Err(RejectResponse {
+            reject_code: RejectCode::CanisterReject,
+            ..
+        })
+    );
+}
+
+fn test_max_number_of_response_headers(env: TestEnv) {
+    let handlers = Handlers::new(&env);
+    let webserver_ipv6 = get_universal_vm_address(&env);
+
+    // HTTP server returns 5 headers in addition to the requested headers.
+    let response_headers = HTTP_HEADERS_MAX_NUMBER - 5;
+    let url = format!(
+        "https://[{}]:20443/{}/{}",
+        webserver_ipv6, "many_response_headers", response_headers
+    );
+
+    let response = block_on(submit_outcall(
+        &handlers,
+        RemoteHttpRequest {
+            request: CanisterHttpRequestArgs {
+                url,
+                headers: BoundedHttpHeaders::new(vec![]),
+                method: HttpMethod::GET,
+                body: None,
+                transform: None,
+                max_response_bytes: None,
+            },
+            cycles: 500_000_000_000,
+        },
+    ))
+    .expect("Request is successful.");
+
+    assert_matches!(&response, RemoteHttpResponse { status: 200, .. });
+    assert_http_response(&response);
```

Comment:
Could we also check that we did receive all those HTTP_HEADERS_MAX_NUMBER headers?